### PR TITLE
Save and restore ymm registers in signal handlers.

### PR DESCRIPTION
**Pull request by adityamandaleeka**
- Modified the CONTEXT structure for storing the upper 16 bytes of ymm registers
- Upon start of signal handler, ymmh data is copied from the native context to the CONTEXT structure, and a new flag (CONTEXT_XSTATE) is set to indicate that the CONTEXT has this information.
- CONTEXTToNativeContext copies the regs back into the native context at the end.

This change fixes only the Linux side for now.

RtlRestoreContext now also checks the XSTATE flag and ymmh data is restored into registers from the CONTEXT structure if it's set. I suppose I can remove this, since we're now returning from the signal handler after doing GCStress without calling RtlRestoreContext. 

@janvorli PTAL

_This won't be merged until it's fully tested._

### CODE REVIEW

File: src/pal/src/arch/i386/asmconstants.h [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 #define CONTEXT_FULL (CONTEXT_CONTROL | CONTEXT_INTEGER | CONTEXT_FLOATING_POINT)
 
+#define CONTEXT_XSTATE 128
```
Comment (janvorli) [inline src/pal/src/arch/i386/asmconstants.h]: This should be 64 to match the value in pal.h

Comment (adityamandaleeka) [inline src/pal/src/arch/i386/asmconstants.h]: Thanks, fixed.

**Comment by janvorli** [comment]
LGTM, although I wonder if we should keep the stuff in the RtlRestoreContext. It is kind of strange when we don't do that in RtlCaptureContext / CONTEXT_CaptureContext.

**Comment by adityamandaleeka** [comment]
Yea, I was thinking the same thing. I'm fine with removing it. It's not hard to add it back later if/when it's needed.

**Comment by janvorli** [comment]
Ok, then let's remove it for now.

**Comment by janvorli** [comment]
LGTM

**Comment by adityamandaleeka** [comment]
@janvorli Done. Do you remember why the following was changed from 10*16? I believe you changed it at some point.

```C++
#define CONTEXT_Xmm0 CONTEXT_FltSave+FLOATING_SAVE_AREA_SIZE // was 10*16
```

This change changes it back to 10*16, so I was wondering whether it was changed for a reason.

**Comment by sergiy-k** [comment]
I believe the code in RtlRestoreContext was needed in @kouvel's original PR because HandleHardwareException was using RtlRestoreContext to resume execution without returning to the signal handler. @janvorli has recently changed HandleHardwareException to not do that. If you are sure that the debugger will not call RtlRestoreContext (in response to a hardware exception/breakpoint) then it should be fine to revert the changes in this function that Kount implemented.

**Comment by adityamandaleeka** [comment]
I added a commit fixing the offset to the VectorRegister data in asmconstants.h. It was previously `CONTEXT_Xmm15+16`, but there is a 96 byte reserved area in `XMM_SAVE_AREA32` after the XMM registers. 

This value appears to have been this way since the first commit to this repo, but I guess it wasn't caught because it wasn't being used before.

**Comment by adityamandaleeka** [comment]
I just added a couple of commits with the following changes:

- I added back the XSTATE stuff in RtlRestoreContext in case someone calls that with a CONTEXT that has CONTEXT_XSTATE set. 
- CONTEXTFromNativeContext will now clear the CONTEXT_XSTATE flag from the context if:
 - If fpregs aren't available in the native context passed in
 - If fpregs is non-null but extended state isn't available in the native context

**Comment by adityamandaleeka** [comment]
@sergiy-k PTAL

**Comment by sergiy-k** [comment]
LGTM

**Comment by adityamandaleeka** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by jasonwilliams200OK** [comment]
@adityamandaleeka, this PR broke Alpine build: https://github.com/dotnet/coreclr/issues/917#issuecomment-236907708. :(

**Pull request merged**